driver-core/src/main/com/mongodb/internal/connection/AwsAuthenticator.java

```diff
@@ -223,20 +223,18 @@ private String getPassword() {
 
         @Nullable
         private String getSessionToken() {
-            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
-            if (credential.getPassword() == null || credential.getUserName() == null) {
-                if (token == null) {
-                    if ((token = System.getenv("AWS_SESSION_TOKEN")) == null) {
-                        token = BsonDocument
-                                .parse(getHttpResponse())
-                                .getString("Token")
-                                .getValue();
-                    }
-                } else {
-                    throw new IllegalArgumentException("The connection string contains auth properties and no username and password");
-                }
+            if (credential.getPassword() != null && credential.getUserName() != null) {
+                return credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
             }
-            return token;
+
+            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null && System.getenv("AWS_ACCESS_KEY_ID") != null) {
+                return System.getenv("AWS_SESSION_TOKEN");
+            }
+
+            return BsonDocument
+                    .parse(getHttpResponse())
+                    .getString("Token")
+                    .getValue();
```

Comment:
@jyemin This fixed a bug with getting creds from env vars. Previously, If the env vars for getting the access key ID and secret access key were set we'd look for a session token in the env var AWS_SESSION_TOKEN. If that env var did not contain a session token then we'd make an HTTP query to retrieve it. That was an issue since a session token may or may not be set in this case. It worked before because the task for checking env vars for creds ran _after_ the task for checking the EC2 endpoint. Therefore, by the time we got to the task for checking env vars for creds the EC2 endpoint was set up and this method retrieved a session token from that endpoint.

Reply:
The EC2 function is now run after all other functions (besides the ECS func) to ensure that none of the other functions use its endpoint.

Reply:
Ack.